File: processor/src/test/java/io/michaelrocks/lightsaber/processor/annotations/proxy/AnnotationProxyGeneratorTest.java
/*
 * Copyright 2015 Michael Rozumyanskiy
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.michaelrocks.lightsaber.processor.annotations.proxy;

import io.michaelrocks.lightsaber.processor.annotations.AnnotationDescriptor;
import io.michaelrocks.lightsaber.processor.annotations.AnnotationDescriptorBuilder;
import org.junit.Before;
import org.junit.Test;
import org.objectweb.asm.Type;

import javax.annotation.ParametersAreNonnullByDefault;
import java.lang.annotation.Annotation;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import java.lang.reflect.Constructor;
import java.lang.reflect.Method;
import java.util.Map;
import java.util.TreeMap;

import static org.junit.Assert.*;

@SuppressWarnings("unused")
@ParametersAreNonnullByDefault
public class AnnotationProxyGeneratorTest {
    @EmptyAnnotation
    @BooleanAnnotation
    @ByteAnnotation
    @CharAnnotation
    @FloatAnnotation
    @DoubleAnnotation
    @IntAnnotation
    @LongAnnotation
    @ShortAnnotation
    @StringAnnotation
    @EnumAnnotation
    @AnnotationAnnotation
    @BooleanArrayAnnotation
    @ByteArrayAnnotation
    @CharArrayAnnotation
    @FloatArrayAnnotation
    @DoubleArrayAnnotation
    @IntArrayAnnotation
    @LongArrayAnnotation
    @ShortArrayAnnotation
    @StringArrayAnnotation
    @EnumArrayAnnotation
    @AnnotationArrayAnnotation
    @CompositeAnnotation
    private static final Object ANNOTATION_HOLDER = new Object();

    private ArrayClassLoader classLoader;

    @Before
    public void createClassLoader() {
        classLoader = new ArrayClassLoader(getClass().getClassLoader());
    }

    @Test
    public void testEqualsByReference() throws Exception {
        final EmptyAnnotation annotation = createAnnotationProxy(EmptyAnnotation.class);
        assertEquals(annotation, annotation);
    }

    @Test
    public void testNotEqualsToNull() throws Exception {
        // noinspection ObjectEqualsNull
        assertFalse(createAnnotationProxy(EmptyAnnotation.class).equals(null));
    }

    @Test
    public void testNotEqualsToOtherType() throws Exception {
        assertNotEquals(new Object(), createAnnotationProxy(EmptyAnnotation.class));
        assertNotEquals(createAnnotationProxy(IntAnnotation.class, 42), createAnnotationProxy(EmptyAnnotation.class));
    }

    @Test
    public void testEmptyAnnotation() throws Exception {
        assertAnnotationEquals(EmptyAnnotation.class);
    }

    @Test
    public void testBooleanAnnotation() throws Exception {
        assertAnnotationEquals(BooleanAnnotation.class, true);
        assertAnnotationNotEquals(BooleanAnnotation.class, false);
    }

    @Test
    public void testByteAnnotation() throws Exception {
        assertAnnotationEquals(ByteAnnotation.class, (byte) 42);
        assertAnnotationNotEquals(ByteAnnotation.class, (byte) -42);
    }

    @Test
    public void testCharAnnotation() throws Exception {
        assertAnnotationEquals(CharAnnotation.class, 'x');
        assertAnnotationNotEquals(CharAnnotation.class, ' ');
    }

    @Test
    public void testFloatAnnotation() throws Exception {
        assertAnnotationEquals(FloatAnnotation.class, 2.7182818284590452354f);
        assertAnnotationNotEquals(FloatAnnotation.class, Float.NaN);
    }

    @Test
    public void testDoubleAnnotation() throws Exception {
        assertAnnotationEquals(DoubleAnnotation.class, 3.14159265358979323846);
        assertAnnotationNotEquals(DoubleAnnotation.class, Double.NaN);
    }

    @Test
    public void testIntAnnotation() throws Exception {
        assertAnnotationEquals(IntAnnotation.class, 42);
        assertAnnotationNotEquals(IntAnnotation.class, -42);
    }

    @Test
    public void testLongAnnotation() throws Exception {
        assertAnnotationEquals(LongAnnotation.class, 42L);
        assertAnnotationNotEquals(LongAnnotation.class, -43L);
    }

    @Test
    public void testShortAnnotation() throws Exception {
        assertAnnotationEquals(ShortAnnotation.class, (short) 42);
        assertAnnotationNotEquals(ShortAnnotation.class, (short) -42);
    }

    @Test
    public void testStringAnnotation() throws Exception {
        assertAnnotationEquals(StringAnnotation.class, "Value");
        assertAnnotationNotEquals(StringAnnotation.class, "Value2");
    }

    @Test
    public void testEnumAnnotation() throws Exception {
        assertAnnotationEquals(EnumAnnotation.class, RetentionPolicy.RUNTIME);
        assertAnnotationNotEquals(EnumAnnotation.class, RetentionPolicy.CLASS);
    }

    @Test
    public void testAnnotationAnnotation() throws Exception {
        assertAnnotationEquals(AnnotationAnnotation.class, createAnnotationProxy(IntAnnotation.class, 42));
        assertAnnotationNotEquals(AnnotationAnnotation.class, createAnnotationProxy(IntAnnotation.class, -42));
    }

    @Test
    public void testBooleanArrayAnnotation() throws Exception {
        assertAnnotationEquals(BooleanArrayAnnotation.class, new Object[] { new boolean[] { true } });
        assertAnnotationNotEquals(BooleanArrayAnnotation.class, new Object[] { new boolean[] { false } });
    }

    @Test
    public void testByteArrayAnnotation() throws Exception {
        assertAnnotationEquals(ByteArrayAnnotation.class, new Object[] { new byte[] { 42 } });
        assertAnnotationNotEquals(ByteArrayAnnotation.class, new Object[] { new byte[] { -42 } });
    }

    @Test
    public void testCharArrayAnnotation() throws Exception {
        assertAnnotationEquals(CharArrayAnnotation.class, new Object[] { new char[] { 'x' } });
        assertAnnotationNotEquals(CharArrayAnnotation.class, new Object[] { new char[] { 'y' } });
    }

    @Test
    public void testFloatArrayAnnotation() throws Exception {
        assertAnnotationEquals(FloatArrayAnnotation.class, new Object[] { new float[] { 2.7182818284590452354f } });
        assertAnnotationNotEquals(FloatArrayAnnotation.class, new Object[] { new float[] { Float.NaN } });
    }

    @Test
    public void testDoubleArrayAnnotation() throws Exception {
        assertAnnotationEquals(DoubleArrayAnnotation.class, new Object[] { new double[] { 3.14159265358979323846 } });
        assertAnnotationNotEquals(DoubleArrayAnnotation.class, new Object[] { new double[] { Double.NaN } });
    }

    @Test
    public void testIntArrayAnnotation() throws Exception {
        assertAnnotationEquals(IntArrayAnnotation.class, new Object[] { new int[] { 42 } });
        assertAnnotationNotEquals(IntArrayAnnotation.class, new Object[] { new int[] { -42 } });
    }

    @Test
    public void testLongArrayAnnotation() throws Exception {
        assertAnnotationEquals(LongArrayAnnotation.class, new Object[] { new long[] { 42L } });
        assertAnnotationNotEquals(LongArrayAnnotation.class, new Object[] { new long[] { -42L } });
    }

    @Test
    public void testShortArrayAnnotation() throws Exception {
        assertAnnotationEquals(ShortArrayAnnotation.class, new Object[] { new short[] { 42 } });
        assertAnnotationNotEquals(ShortArrayAnnotation.class, new Object[] { new short[] { -42 } });
    }

    @Test
    public void testStringArrayAnnotation() throws Exception {
        assertAnnotationEquals(StringArrayAnnotation.class, new Object[] { new String[] { "Value" } });
        assertAnnotationNotEquals(StringArrayAnnotation.class, new Object[] { new String[] { "Value2" } });
    }

    @Test
    public void testEnumArrayAnnotation() throws Exception {
        assertAnnotationEquals(EnumArrayAnnotation.class,
                new Object[] { new RetentionPolicy[] { RetentionPolicy.RUNTIME } });
        assertAnnotationNotEquals(EnumArrayAnnotation.class,
                new Object[] { new RetentionPolicy[] { RetentionPolicy.CLASS } });
    }

    @Test
    public void testAnnotationArrayAnnotation() throws Exception {
        assertAnnotationEquals(AnnotationArrayAnnotation.class,
                new Object[] { new IntAnnotation[] { createAnnotationProxy(IntAnnotation.class, 42) } });
        assertAnnotationNotEquals(AnnotationArrayAnnotation.class,
                new Object[] { new IntAnnotation[] { createAnnotationProxy(IntAnnotation.class, -42) } });
    }

    @Test
    public void testCompositeAnnotation() throws Exception {
        assertAnnotationEquals(CompositeAnnotation.class,
                true,
                (byte) 42,
                'x',
                2.7182818284590452354f,
                3.14159265358979323846,
                42,
                42L,
                (short) 42,
                "Value",
                RetentionPolicy.RUNTIME,
                createAnnotationProxy(IntAnnotation.class, 42),
                new boolean[] { true },
                new byte[] { (byte) 42 },
                new char[] { 'x' },
                new float[] { 2.7182818284590452354f },
                new double[] { 3.14159265358979323846 },
                new int[] { 42 },
                new long[] { 42L },
                new short[] { (short) 42 },
                new String[] { "Value" },
                new RetentionPolicy[] { RetentionPolicy.RUNTIME },
                new IntAnnotation[] { createAnnotationProxy(IntAnnotation.class, 42) });
    }

    @Test
    public void testEmptyToString() throws Exception {
        final EmptyAnnotation annotation = createAnnotationProxy(EmptyAnnotation.class);
        final String expected = "@" + EmptyAnnotation.class.getName() + "()";
        final String actual = annotation.toString();
        assertEquals(expected, actual);
    }

    @Test
    public void testCompositeToString() throws Exception {
        final CompositeAnnotation annotation = createAnnotationProxy(CompositeAnnotation.class,
                true,
                (byte) 42,
                'x',
                Float.NaN,
                Double.POSITIVE_INFINITY,
                42,
                42L,
                (short) 42,
                "Value",
                RetentionPolicy.RUNTIME,
                createAnnotationProxy(IntAnnotation.class, 42),
                new boolean[] { true, false },
                new byte[] { 42, 43 },
                new char[] { 'x', 'y' },
                new float[] { Float.NaN, Float.POSITIVE_INFINITY },
                new double[] { Double.POSITIVE_INFINITY, Double.NaN },
                new int[] { 42, 43 },
                new long[] { 42L, 43L },
                new short[] { 42, 43 },
                new String[] { "Value1", "Value2" },
                new RetentionPolicy[] { RetentionPolicy.RUNTIME, RetentionPolicy.CLASS },
                new IntAnnotation[] {
                        createAnnotationProxy(IntAnnotation.class, 42),
                        createAnnotationProxy(IntAnnotation.class, 43)
                });

        final String expected = "@" + CompositeAnnotation.class.getName() + "("
                + "booleanValue=true, "
                + "byteValue=42, "
                + "charValue=x, "
                + "floatValue=NaN, "
                + "doubleValue=Infinity, "
                + "intValue=42, "
                + "longValue=42, "
                + "shortValue=42, "
                + "stringValue=Value, "
                + "enumValue=RUNTIME, "
                + "annotationValue=@" + IntAnnotation.class.getName() + "(value=42), "
                + "booleanArrayValue=[true, false], "
                + "byteArrayValue=[42, 43], "
                + "charArrayValue=[x, y], "
                + "floatArrayValue=[NaN, Infinity], "
                + "doubleArrayValue=[Infinity, NaN], "
                + "intArrayValue=[42, 43], "
                + "longArrayValue=[42, 43], "
                + "shortArrayValue=[42, 43], "
                + "stringArrayValue=[Value1, Value2], "
                + "enumArrayValue=[RUNTIME, CLASS], "
                + "annotationArrayValue=["
                + "@" + IntAnnotation.class.getName() + "(value=42), @" + IntAnnotation.class.getName() + "(value=43)]"
                + ")";
        final String actual = annotation.toString();
        assertEquals(expected, actual);
    }

    private <T extends Annotation> void assertAnnotationEquals(final Class<T> annotationClass,
            final Object... values) throws Exception {
        final T expectedAnnotation = getAnnotation(annotationClass);
        final T actualAnnotation = createAnnotationProxy(annotationClass, values);
        assertEquals(expectedAnnotation, actualAnnotation);
        assertEquals(expectedAnnotation.hashCode(), actualAnnotation.hashCode());
        assertEquals(expectedAnnotation.annotationType(), actualAnnotation.annotationType());
    }

    private <T extends Annotation> void assertAnnotationNotEquals(final Class<T> annotationClass,
            final Object... values) throws Exception {
        final T expectedAnnotation = getAnnotation(annotationClass);
        final T actualAnnotation = createAnnotationProxy(annotationClass, values);
        assertNotEquals(expectedAnnotation, actualAnnotation);
        assertEquals(expectedAnnotation.annotationType(), actualAnnotation.annotationType());
    }

    private <T extends Annotation> T createAnnotationProxy(final Class<T> annotationClass,
            final Object... values) throws Exception {
        final String annotationProxyClassName = getAnnotationProxyClassName(annotationClass);
        if (!classLoader.hasClass(annotationProxyClassName)) {
            addAnnotationProxy(annotationClass);
        }
        // noinspection unchecked
        final Class<T> annotationProxyClass = (Class<T>) classLoader.loadClass(annotationProxyClassName);
        assertEquals("Annotation proxy must have the only constructor",
                1, annotationProxyClass.getDeclaredConstructors().length);
        // noinspection unchecked
        final Constructor<T> constructor = (Constructor<T>) annotationProxyClass.getDeclaredConstructors()[0];
        return constructor.newInstance(values);
    }

    private void addAnnotationProxy(final Class<? extends Annotation> annotationClass) {
        final AnnotationDescriptor annotationDescriptor = getAnnotationDescriptor(annotationClass);
        final Type annotationProxyType = Type.getObjectType(getAnnotationProxyClassName(annotationClass));
        final AnnotationProxyGenerator annotationProxyGenerator =
                new AnnotationProxyGenerator(annotationDescriptor, annotationProxyType);
        classLoader.addClass(annotationProxyType.getInternalName(), annotationProxyGenerator.generate());
    }

    private static String getAnnotationProxyClassName(final Class<? extends Annotation> annotationClass) {
        return annotationClass.getSimpleName() + "Proxy";
    }

    private static AnnotationDescriptor getAnnotationDescriptor(final Class<? extends Annotation> annotationClass) {
        final Map<Integer, Method> orderedMethods = new TreeMap<>();
        for (final Method method : annotationClass.getDeclaredMethods()) {
            final Order orderAnnotation = method.getAnnotation(Order.class);
            final int order = orderAnnotation == null ? 0 : orderAnnotation.value();
            final Method oldMethod = orderedMethods.put(order, method);
            assertNull("Method order must be distinct", oldMethod);
        }

        final Type annotationType = Type.getType(annotationClass);
        final AnnotationDescriptorBuilder builder = new AnnotationDescriptorBuilder(annotationType);
        for (final Method method : orderedMethods.values()) {
            builder.addField(method.getName(), Type.getType(method.getReturnType()));
        }
        return builder.build();
    }

    private <T extends Annotation> T getAnnotation(final Class<T> annotationClass) throws Exception {
        for (final Annotation annotation : getClass().getDeclaredField("ANNOTATION_HOLDER").getAnnotations()) {
            if (annotation.annotationType().equals(annotationClass)) {
                return annotationClass.cast(annotation);
            }
        }
        fail("Annotation for class " + annotationClass + " not found");
        throw new RuntimeException();
    }

    @Retention(RetentionPolicy.RUNTIME)
    public @interface EmptyAnnotation {
    }

    @Retention(RetentionPolicy.RUNTIME)
    public @interface BooleanAnnotation {
        boolean value() default true;
    }

    @Retention(RetentionPolicy.RUNTIME)
    public @interface ByteAnnotation {
        byte value() default 42;
    }

    @Retention(RetentionPolicy.RUNTIME)
    public @interface CharAnnotation {
        char value() default 'x';
    }

    @Retention(RetentionPolicy.RUNTIME)
    public @interface FloatAnnotation {
        float value() default 2.7182818284590452354f;
    }

    @Retention(RetentionPolicy.RUNTIME)
    public @interface DoubleAnnotation {
        double value() default 3.14159265358979323846;
    }

    @Retention(RetentionPolicy.RUNTIME)
    public @interface IntAnnotation {
        int value() default 42;
    }

    @Retention(RetentionPolicy.RUNTIME)
    public @interface LongAnnotation {
        long value() default 42L;
    }

    @Retention(RetentionPolicy.RUNTIME)
    public @interface ShortAnnotation {
        short value() default 42;
    }

    @Retention(RetentionPolicy.RUNTIME)
    public @interface StringAnnotation {
        String value() default "Value";
    }

    @Retention(RetentionPolicy.RUNTIME)
    public @interface EnumAnnotation {
        RetentionPolicy value() default RetentionPolicy.RUNTIME;
    }

    @Retention(RetentionPolicy.RUNTIME)
    public @interface AnnotationAnnotation {
        IntAnnotation value() default @IntAnnotation;
    }

    @Retention(RetentionPolicy.RUNTIME)
    public @interface BooleanArrayAnnotation {
        boolean[] value() default { true };
    }

    @Retention(RetentionPolicy.RUNTIME)
    public @interface ByteArrayAnnotation {
        byte[] value() default { 42 };
    }

    @Retention(RetentionPolicy.RUNTIME)
    public @interface CharArrayAnnotation {
        char[] value() default { 'x' };
    }

    @Retention(RetentionPolicy.RUNTIME)
    public @interface FloatArrayAnnotation {
        float[] value() default { 2.7182818284590452354f };
    }

    @Retention(RetentionPolicy.RUNTIME)
    public @interface DoubleArrayAnnotation {
        double[] value() default { 3.14159265358979323846 };
    }

    @Retention(RetentionPolicy.RUNTIME)
    public @interface IntArrayAnnotation {
        int[] value() default { 42 };
    }

    @Retention(RetentionPolicy.RUNTIME)
    public @interface LongArrayAnnotation {
        long[] value() default { 42L };
    }

    @Retention(RetentionPolicy.RUNTIME)
    public @interface ShortArrayAnnotation {
        short[] value() default { 42 };
    }

    @Retention(RetentionPolicy.RUNTIME)
    public @interface StringArrayAnnotation {
        String[] value() default { "Value" };
    }

    @Retention(RetentionPolicy.RUNTIME)
    public @interface EnumArrayAnnotation {
        RetentionPolicy[] value() default { RetentionPolicy.RUNTIME };
    }

    @Retention(RetentionPolicy.RUNTIME)
    public @interface AnnotationArrayAnnotation {
        IntAnnotation[] value() default { @IntAnnotation };
    }

    @Retention(RetentionPolicy.RUNTIME)
    public @interface CompositeAnnotation {
        @Order( 0) boolean booleanValue() default true;
        @Order( 1) byte byteValue() default 42;
        @Order( 2) char charValue() default 'x';
        @Order( 3) float floatValue() default 2.7182818284590452354f;
        @Order( 4) double doubleValue() default 3.14159265358979323846;
        @Order( 5) int intValue() default 42;
        @Order( 6) long longValue() default 42L;
        @Order( 7) short shortValue() default 42;
        @Order( 8) String stringValue() default "Value";
        @Order( 9) RetentionPolicy enumValue() default RetentionPolicy.RUNTIME;
        @Order(10) IntAnnotation annotationValue() default @IntAnnotation;
        @Order(11) boolean[] booleanArrayValue() default { true };
        @Order(12) byte[] byteArrayValue() default { 42 };
        @Order(13) char[] charArrayValue() default { 'x' };
        @Order(14) float[] floatArrayValue() default { 2.7182818284590452354f };
        @Order(15) double[] doubleArrayValue() default { 3.14159265358979323846 };
        @Order(16) int[] intArrayValue() default { 42 };
        @Order(17) long[] longArrayValue() default { 42L };
        @Order(18) short[] shortArrayValue() default { 42 };
        @Order(19) String[] stringArrayValue() default { "Value" };
        @Order(20) RetentionPolicy[] enumArrayValue() default { RetentionPolicy.RUNTIME };
        @Order(21) IntAnnotation[] annotationArrayValue() default { @IntAnnotation };
    }

    @Retention(RetentionPolicy.RUNTIME)
    @Target(ElementType.METHOD)
    private @interface Order {
        int value();
    }
}

File: processor/src/test/kotlin/io/michaelrocks/lightsaber/processor/annotations/proxy/AnnotationProxyGeneratorTest.kt
/*
 * Copyright 2016 Michael Rozumyanskiy
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.michaelrocks.lightsaber.processor.annotations.proxy

import io.michaelrocks.grip.ClassRegistry
import io.michaelrocks.grip.mirrors.ClassMirror
import io.michaelrocks.grip.mirrors.MethodMirror
import io.michaelrocks.lightsaber.mockito.*
import io.michaelrocks.lightsaber.processor.annotations.proxy.Annotations.*
import io.michaelrocks.lightsaber.processor.commons.Types
import io.michaelrocks.lightsaber.processor.commons.internalName
import io.michaelrocks.lightsaber.processor.commons.type
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import org.objectweb.asm.Opcodes
import org.objectweb.asm.Type
import java.lang.annotation.RetentionPolicy
import java.lang.reflect.Constructor
import java.lang.reflect.InvocationHandler
import java.lang.reflect.Method
import java.lang.reflect.Proxy
import java.util.*

class AnnotationProxyGeneratorTest {
  companion object {
    @field:EmptyAnnotation
    @field:BooleanAnnotation
    @field:ByteAnnotation
    @field:CharAnnotation
    @field:FloatAnnotation
    @field:DoubleAnnotation
    @field:IntAnnotation
    @field:LongAnnotation
    @field:ShortAnnotation
    @field:StringAnnotation
    @field:EnumAnnotation
    @field:ClassAnnotation
    @field:AnnotationAnnotation
    @field:BooleanArrayAnnotation
    @field:ByteArrayAnnotation
    @field:CharArrayAnnotation
    @field:FloatArrayAnnotation
    @field:DoubleArrayAnnotation
    @field:IntArrayAnnotation
    @field:LongArrayAnnotation
    @field:ShortArrayAnnotation
    @field:StringArrayAnnotation
    @field:EnumArrayAnnotation
    @field:ClassArrayAnnotation
    @field:AnnotationArrayAnnotation
    @field:CompositeAnnotation
    private val ANNOTATION_HOLDER = Object()

    private fun getAnnotationProxyClassName(annotationClass: Class<out Annotation>): String {
      return annotationClass.simpleName + "Proxy"
    }

    private fun getAnnotationDescriptor(annotationClass: Class<out Annotation>): ClassMirror {
      val orderedMethods = TreeMap<Int, Method>()
      for (method in annotationClass.declaredMethods) {
        val orderAnnotation = method.getAnnotation(Order::class.java)
        val order = if (orderAnnotation == null) 0 else orderAnnotation.value
        val oldMethod = orderedMethods.put(order, method)
        assertNull("Method order must be distinct", oldMethod)
      }

      return ClassMirror.Builder()
          .name(annotationClass.internalName)
          .run {
            for (method in orderedMethods.values) {
              addMethod(
                  MethodMirror.Builder()
                      .name(method.name)
                      .type(Type.getMethodType(method.returnType.type))
                      .build()
              )
            }
            build()
          }
    }

    private fun <T : Annotation> getAnnotation(annotationClass: Class<T>): T {
      for (annotation in AnnotationProxyGeneratorTest::class.java.getDeclaredField("ANNOTATION_HOLDER").annotations) {
        if (annotation.annotationClass.java == annotationClass) {
          return annotationClass.cast(annotation)
        }
      }
      fail("Annotation for class $annotationClass not found")
      throw RuntimeException()
    }
  }

  private lateinit var classLoader: ArrayClassLoader

  @Before
  fun createClassLoader() {
    val annotations = Annotations()
    classLoader = ArrayClassLoader(annotations.javaClass.classLoader)
  }

  @Test
  fun testEqualsByReference() {
    val annotation = createAnnotationProxy<EmptyAnnotation>()
    assertEquals(annotation, annotation)
  }

  @Test
  fun testNotEqualsToNull() {
    @Suppress("SENSELESS_COMPARISON")
    assertFalse(createAnnotationProxy<EmptyAnnotation>() == null)
  }

  @Test
  fun testNotEqualsToOtherType() {
    assertNotEquals(Object(), createAnnotationProxy<EmptyAnnotation>())
    assertNotEquals(createAnnotationProxy<IntAnnotation>(42), createAnnotationProxy<EmptyAnnotation>())
  }

  @Test
  fun testEmptyAnnotation() {
    assertAnnotationEquals<EmptyAnnotation>()
  }

  @Test
  fun testBooleanAnnotation() {
    assertAnnotationEquals<BooleanAnnotation>(true)
    assertAnnotationNotEquals<BooleanAnnotation>(false)
  }

  @Test
  fun testByteAnnotation() {
    assertAnnotationEquals<ByteAnnotation>(42.toByte())
    assertAnnotationNotEquals<ByteAnnotation>((-42).toByte())
  }

  @Test
  fun testCharAnnotation() {
    assertAnnotationEquals<CharAnnotation>('x')
    assertAnnotationNotEquals<CharAnnotation>(' ')
  }

  @Test
  fun testFloatAnnotation() {
    assertAnnotationEquals<FloatAnnotation>(2.7182818284590452354f)
    assertAnnotationNotEquals<FloatAnnotation>(Float.NaN)
  }

  @Test
  fun testDoubleAnnotation() {
    assertAnnotationEquals<DoubleAnnotation>(3.14159265358979323846)
    assertAnnotationNotEquals<DoubleAnnotation>(Double.NaN)
  }

  @Test
  fun testIntAnnotation() {
    assertAnnotationEquals<IntAnnotation>(42)
    assertAnnotationNotEquals<IntAnnotation>(-42)
  }

  @Test
  fun testLongAnnotation() {
    assertAnnotationEquals<LongAnnotation>(42L)
    assertAnnotationNotEquals<LongAnnotation>(-43L)
  }

  @Test
  fun testShortAnnotation() {
    assertAnnotationEquals<ShortAnnotation>(42.toShort())
    assertAnnotationNotEquals<ShortAnnotation>((-42).toShort())
  }

  @Test
  fun testStringAnnotation() {
    assertAnnotationEquals<StringAnnotation>("Value")
    assertAnnotationNotEquals<StringAnnotation>("Value2")
  }

  @Test
  fun testEnumAnnotation() {
    assertAnnotationEquals<EnumAnnotation>(RetentionPolicy.RUNTIME)
    assertAnnotationNotEquals<EnumAnnotation>(RetentionPolicy.CLASS)
  }

  @Test
  fun testClassAnnotation() {
    assertAnnotationEquals<ClassAnnotation>(Any::class.java)
    assertAnnotationNotEquals<ClassAnnotation>(String::class.java)
  }

  @Test
  fun testAnnotationAnnotation() {
    assertAnnotationEquals<AnnotationAnnotation>(createAnnotationProxy<IntAnnotation>(42))
    assertAnnotationNotEquals<AnnotationAnnotation>(createAnnotationProxy<IntAnnotation>(-42))
  }

  @Test
  fun testBooleanArrayAnnotation() {
    assertAnnotationEquals<BooleanArrayAnnotation>(booleanArrayOf(true))
    assertAnnotationNotEquals<BooleanArrayAnnotation>(booleanArrayOf(false))
  }

  @Test
  fun testByteArrayAnnotation() {
    assertAnnotationEquals<ByteArrayAnnotation>(byteArrayOf(42))
    assertAnnotationNotEquals<ByteArrayAnnotation>(byteArrayOf(-42))
  }

  @Test
  fun testCharArrayAnnotation() {
    assertAnnotationEquals<CharArrayAnnotation>(charArrayOf('x'))
    assertAnnotationNotEquals<CharArrayAnnotation>(charArrayOf('y'))
  }

  @Test
  fun testFloatArrayAnnotation() {
    assertAnnotationEquals<FloatArrayAnnotation>(floatArrayOf(2.7182818284590452354f))
    assertAnnotationNotEquals<FloatArrayAnnotation>(floatArrayOf(Float.NaN))
  }

  @Test
  fun testDoubleArrayAnnotation() {
    assertAnnotationEquals<DoubleArrayAnnotation>(doubleArrayOf(3.14159265358979323846))
    assertAnnotationNotEquals<DoubleArrayAnnotation>(doubleArrayOf(Double.NaN))
  }

  @Test
  fun testIntArrayAnnotation() {
    assertAnnotationEquals<IntArrayAnnotation>(intArrayOf(42))
    assertAnnotationNotEquals<IntArrayAnnotation>(intArrayOf(-42))
  }

  @Test
  fun testLongArrayAnnotation() {
    assertAnnotationEquals<LongArrayAnnotation>(longArrayOf(42L))
    assertAnnotationNotEquals<LongArrayAnnotation>(longArrayOf(-42L))
  }

  @Test
  fun testShortArrayAnnotation() {
    assertAnnotationEquals<ShortArrayAnnotation>(shortArrayOf(42))
    assertAnnotationNotEquals<ShortArrayAnnotation>(shortArrayOf(-42))
  }

  @Test
  fun testStringArrayAnnotation() {
    assertAnnotationEquals<StringArrayAnnotation>(arrayOf("Value"))
    assertAnnotationNotEquals<StringArrayAnnotation>(arrayOf("Value2"))
  }

  @Test
  fun testEnumArrayAnnotation() {
    assertAnnotationEquals<EnumArrayAnnotation>(arrayOf(RetentionPolicy.RUNTIME))
    assertAnnotationNotEquals<EnumArrayAnnotation>(arrayOf(RetentionPolicy.CLASS))
  }

  @Test
  fun testClassArrayAnnotation() {
    assertAnnotationEquals<ClassArrayAnnotation>(arrayOf<Class<*>>(Any::class.java))
    assertAnnotationNotEquals<ClassArrayAnnotation>(arrayOf<Class<*>>(String::class.java))
  }

  @Test
  fun testAnnotationArrayAnnotation() {
    assertAnnotationEquals<AnnotationArrayAnnotation>(arrayOf(createAnnotationProxy<IntAnnotation>(42)))
    assertAnnotationNotEquals<AnnotationArrayAnnotation>(arrayOf(createAnnotationProxy<IntAnnotation>(-42)))
  }

  @Test
  fun testCompositeAnnotation() {
    assertAnnotationEquals<CompositeAnnotation>(
        true,
        42.toByte(),
        'x',
        2.7182818284590452354f,
        3.14159265358979323846,
        42,
        42L,
        42.toShort(),
        "Value",
        RetentionPolicy.RUNTIME,
        Any::class.java,
        createAnnotationProxy<IntAnnotation>(42),
        booleanArrayOf(true),
        byteArrayOf(42.toByte()),
        charArrayOf('x'),
        floatArrayOf(2.7182818284590452354f),
        doubleArrayOf(3.14159265358979323846),
        intArrayOf(42),
        longArrayOf(42L),
        shortArrayOf(42.toShort()),
        arrayOf("Value"),
        arrayOf(RetentionPolicy.RUNTIME),
        arrayOf<Class<*>>(Any::class.java),
        arrayOf(createAnnotationProxy<IntAnnotation>(42)))
  }

  @Test
  fun testEmptyToString() {
    val annotation = createAnnotationProxy<EmptyAnnotation>()
    val expected = "@${EmptyAnnotation::class.java.name}()"
    val actual = annotation.toString()
    assertEquals(expected, actual)
  }

  @Test
  fun testCompositeToString() {
    val annotation = createAnnotationProxy<CompositeAnnotation>(
        true,
        42.toByte(),
        'x',
        Float.NaN,
        Double.POSITIVE_INFINITY,
        42,
        42L,
        42.toShort(),
        "Value",
        RetentionPolicy.RUNTIME,
        Any::class.java,
        createAnnotationProxy<IntAnnotation>(42),
        booleanArrayOf(true, false),
        byteArrayOf(42, 43),
        charArrayOf('x', 'y'),
        floatArrayOf(Float.NaN, Float.POSITIVE_INFINITY),
        doubleArrayOf(Double.POSITIVE_INFINITY, Double.NaN),
        intArrayOf(42, 43),
        longArrayOf(42L, 43L),
        shortArrayOf(42, 43),
        arrayOf("Value1", "Value2"),
        arrayOf(RetentionPolicy.RUNTIME, RetentionPolicy.CLASS),
        arrayOf<Class<*>>(Any::class.java),
        arrayOf(createAnnotationProxy<IntAnnotation>(42), createAnnotationProxy<IntAnnotation>(43)))

    val expected = "@${CompositeAnnotation::class.java.name}(" +
        "booleanValue=true, " +
        "byteValue=42, " +
        "charValue=x, " +
        "floatValue=NaN, " +
        "doubleValue=Infinity, " +
        "intValue=42, " +
        "longValue=42, " +
        "shortValue=42, " +
        "stringValue=Value, " +
        "enumValue=RUNTIME, " +
        "classValue=class java.lang.Object, " +
        "annotationValue=@${IntAnnotation::class.java.name}(value=42), " +
        "booleanArrayValue=[true, false], " +
        "byteArrayValue=[42, 43], " +
        "charArrayValue=[x, y], " +
        "floatArrayValue=[NaN, Infinity], " +
        "doubleArrayValue=[Infinity, NaN], " +
        "intArrayValue=[42, 43], " +
        "longArrayValue=[42, 43], " +
        "shortArrayValue=[42, 43], " +
        "stringArrayValue=[Value1, Value2], " +
        "enumArrayValue=[RUNTIME, CLASS], " +
        "classArrayValue=[class java.lang.Object], " +
        "annotationArrayValue=[" +
        "@${IntAnnotation::class.java.name}(value=42), @${IntAnnotation::class.java.name}(value=43)]" +
        ")"
    val actual = annotation.toString()
    assertEquals(expected, actual)
  }

  @Test
  fun testHashCodeCaching() {
    val handler = mock<InvocationHandler>()
    val intAnnotation = newProxyInstance<IntAnnotation>(handler)
    val hashCodeMethod = Any::class.java.getDeclaredMethod("hashCode")
    given(handler.invoke(intAnnotation, hashCodeMethod, null)).thenReturn(0)

    val annotation = createAnnotationProxy<AnnotationAnnotation>(intAnnotation)
    annotation.hashCode()
    annotation.hashCode()
    verify(handler, times(1)).invoke(intAnnotation, hashCodeMethod, null)
    verifyNoMoreInteractions(handler)
  }

  @Test
  fun testToStringCaching() {
    val handler = mock<InvocationHandler>()
    val intAnnotation = newProxyInstance<IntAnnotation>(handler)
    val toStringMethod = Any::class.java.getDeclaredMethod("toString")
    given(handler.invoke(intAnnotation, toStringMethod, null)).thenReturn("ToString")

    val annotation = createAnnotationProxy<AnnotationAnnotation>(intAnnotation)
    annotation.toString()
    annotation.toString()
    verify(handler, times(1)).invoke(intAnnotation, toStringMethod, null)
    verifyNoMoreInteractions(handler)
  }

  private inline fun <reified T : Annotation> assertAnnotationEquals(vararg values: Any) {
    val annotationClass = T::class.java
    val expectedAnnotation = getAnnotation(annotationClass)
    val actualAnnotation = createAnnotationProxy<T>(*values)
    assertEquals(expectedAnnotation, actualAnnotation)
    assertEquals(expectedAnnotation.hashCode().toLong(), actualAnnotation.hashCode().toLong())
    assertEquals(expectedAnnotation.annotationClass.java, actualAnnotation.annotationClass.java)
  }

  private inline fun <reified T : Annotation> assertAnnotationNotEquals(vararg values: Any) {
    val annotationClass = T::class.java
    val expectedAnnotation = getAnnotation(annotationClass)
    val actualAnnotation = createAnnotationProxy<T>(*values)
    assertNotEquals(expectedAnnotation, actualAnnotation)
    assertEquals(expectedAnnotation.annotationClass.java, actualAnnotation.annotationClass.java)
  }

  private inline fun <reified T : Annotation> createAnnotationProxy(vararg values: Any): T {
    val annotationClass = T::class.java
    val annotationProxyClassName = getAnnotationProxyClassName(annotationClass)
    if (!classLoader.hasClass(annotationProxyClassName)) {
      addAnnotationProxy(annotationClass)
    }
    @Suppress("UNCHECKED_CAST")
    val annotationProxyClass = classLoader.loadClass(annotationProxyClassName) as Class<T>
    assertEquals("Annotation proxy must have the only constructor",
        1, annotationProxyClass.declaredConstructors.size.toLong())
    @Suppress("UNCHECKED_CAST")
    val constructor = annotationProxyClass.declaredConstructors[0] as Constructor<T>
    return constructor.newInstance(*values)
  }

  private fun addAnnotationProxy(annotationClass: Class<out Annotation>) {
    val classRegistry = mock<ClassRegistry>()
    given(classRegistry.getClassMirror(notNull())).thenAnswer { invocation ->
      mock<ClassMirror>().apply {
        given(access).thenReturn(Opcodes.ACC_PUBLIC or Opcodes.ACC_SUPER)
        given(type).thenReturn(invocation.arguments[0] as Type)
        given(superType).thenReturn(Types.OBJECT_TYPE)
      }
    }
    val annotationDescriptor = getAnnotationDescriptor(annotationClass)
    val annotationProxyType = Type.getObjectType(getAnnotationProxyClassName(annotationClass))
    val annotationProxyGenerator = AnnotationProxyGenerator(classRegistry, annotationDescriptor, annotationProxyType)
    classLoader.addClass(annotationProxyType.internalName, annotationProxyGenerator.generate())
  }

  private inline fun <reified T : Any> newProxyInstance(handler: InvocationHandler): T =
      Proxy.newProxyInstance(T::class.java.classLoader, arrayOf(T::class.java), handler) as T
}
